### PR TITLE
SDK-binding support for Blob Storage worker extension

### LINKED ISSUE #1760: Release Blob Storage worker extension with SDK-binding support
- [ ]  Merge blob worker extension sdk-binding feature in to main
- [ ]  Release stable blob worker extension package

### PR DESCRIPTION
**Pull request by liliankasem**
### Issue describing the changes in this PR

resolves #1760

### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
  * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+                {
+                    return ConversionResult.Failed(new InvalidOperationException($"Unable to convert blob binding data to type '{context.TargetType.Name}'."));
+                }
+
+                return ConversionResult.Success(result);
+            }
+            catch (JsonException ex)
+            {
+                string msg = string.Format(CultureInfo.CurrentCulture,
+                    @"Binding parameters to complex objects uses JSON serialization.
```
Comment (aishwaryabh) [inline extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs:75]: IIRC this string is used for other converters too. It might be useful to have this as a const in a shared file to standardize the message

File: extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+                result.CopyTo(arrayResult, 0);
+                return arrayResult;
+            }
+
+            return result;
+        }
+
+        private async Task<object?> ToTargetTypeAsync(Type targetType, BlobContainerClient containerClient, string blobName) => targetType switch
+        {
+            Type _ when targetType == typeof(BlobContainerClient) => containerClient,
```
Comment (aishwaryabh) [inline extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs:211]: do we wanna create separate converters for each target type? This can definitely be done in another PR too since this one's already so big

Comment (liliankasem) [reply extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs:211]: We have an issue tracking this as an post-GA item

File: extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs (hunk 1 above)
Comment (liliankasem) [reply extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs:75]: There are a couple constants improvements we can make so I'll leave that for a new PR

File: test/E2ETests/E2ETests/Storage/QueueEndToEndTests.cs [hunk 3]
```diff
@@ -166,60 +166,5 @@ public async Task QueueOutput_PocoList_Succeeds()
             IEnumerable<string> queueMessages = await StorageHelpers.ReadMessagesFromQueue(Constants.Queue.OutputBindingNamePOCO);
             Assert.True(queueMessages.All(msg => msg.Contains(expectedQueueMessage)));
         }
-
-        [Fact]
```
Comment (aishwaryabh) [inline test/E2ETests/E2ETests/Storage/QueueEndToEndTests.cs:170]: Are these tests being put elsewhere?

**Review by aishwaryabh — APPROVED** [review]
LGTM!

File: test/E2ETests/E2ETests/Storage/QueueEndToEndTests.cs (hunk 3 above)
Comment (liliankasem) [reply test/E2ETests/E2ETests/Storage/QueueEndToEndTests.cs:170]: Yeah they were moved to the BlobEndToEndTests file

File: extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+using Microsoft.Azure.Functions.Worker.Extensions.Storage.Blobs;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.Functions.Worker
+{
+    /// <summary>
+    /// Converter to bind Blob Storage type parameters.
+    /// </summary>
+    [SupportsDeferredBinding]
```
Comment (fabiocav) [inline extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs:29]: Can you confirm this converter has no changes from the review done last time before checking it into the feature branch?

File: extensions/Worker.Extensions.Shared/Exceptions/InvalidContentTypeException.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    /// <summary>
+    /// The exception that is thrown when an invalid content-type is provided.
+    /// </summary>
+    internal class InvalidContentTypeException : InvalidOperationException
+    {
+        /// <summary>
+        /// Initializes a new instance of the InvalidContentTypeException class with a specified error message.
+        /// </summary>
+        /// <param name="contentType">The content type(s) that is supported.</param>
+        public InvalidContentTypeException(string contentType) : base($"Unexpected content-type. Only '{contentType}' is supported.") { }
```
Comment (fabiocav) [inline extensions/Worker.Extensions.Shared/Exceptions/InvalidContentTypeException.cs:17]: Same as with the source comment above, it would be helpful to pass the "actual" content type used.

File: extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs (hunk 4 above)
Comment (liliankasem) [reply extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs:29]: Yup, no changes between this and feature branch

Comment (liliankasem) [reply extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs:29]: no there is a change since I updated the exceptions to include the actual source/content-type

**Review by jviau — APPROVED** [review]

**Review by fabiocav — APPROVED** [review]
